iceoryx_hoofs/vocabulary/include/iox/semantic_string.hpp

```diff
@@ -0,0 +1,211 @@
+///     using Parent::Parent;
+/// };
+/// @endcode
+/// @note Since the inner logic of the SemanticString is always the same additional
+///         implementations can be verified directly by the test suite defined in
+///         'test_vocabulary_semantic_string.cpp'.
+///         One has to only add the specific implementation to the 'Implementations'
+///         type list.
+template <typename Child,
+          uint64_t Capacity,
+          DoesContainInvalidContent<Capacity> DoesContainInvalidContentCall,
+          DoesContainInvalidCharacter<Capacity> DoesContainInvalidCharacterCall>
+class SemanticString
+{
+  public:
+    /// @brief Creates a new SemanticString from the provided string literal.
+    ///         If the value contains invalid characters or invalid content
+    ///         the expected returns an error describing the cause.
+    /// @param[in] value the value of the SemanticString
+    /// @return expected either containing the new SemanticString or an error
+    template <uint64_t N>
+    // avoid-c-arrays: we would like to assign string_literals, safe since it is known
+    //                 at compile time.
+    // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays, hicpp-explicit-conversions)
+    static expected<Child, SemanticStringError> create(const char (&value)[N]) noexcept;
+
+    /// @brief Creates a new SemanticString from the provided string.
+    ///         If the value contains invalid characters or invalid content
+    ///         the expected returns an error describing the cause.
+    /// @param[in] value the value of the SemanticString
+    /// @return expected either containing the new SemanticString or an error
+    template <uint64_t N>
+    static expected<Child, SemanticStringError> create(const string<N>& value) noexcept;
+
+    /// @brief Returns the number of characters.
+    /// @return number of characters
+    constexpr uint64_t size() const noexcept;
+
+    /// @brief Returns the capacity of the string.
+    /// @return the maximum amount of characters which can be stored.
+    static constexpr uint64_t capacity() noexcept;
+
+    /// @brief Returns a const reference to the underlying string. It is const
+    ///         and shall not be modified to guarantee the contract that a
+    ///         SemanticString contains always a valid value.
+    /// @return string reference containing the actual value.
+    constexpr const string<Capacity>& as_string() const noexcept;
+
+    /// @brief Appends another string to the SemanticString. If the value contains
+    ///        invalid characters or the result would end up in invalid content
+    ///        it fails.
+    /// @param[in] value the value which should be added
+    /// @return on failure the error inside the expected describes the failure
+    template <typename T>
+    expected<SemanticStringError> append(const T& value) noexcept;
```

Comment:
I wonder if we should restrict `T` here via `IsStringOrCharArray`.

Reply:
I could do this but then the code would look ugly and it is nevertheless handled by the underlying string.
When I solve this at this level the compiler failures are also not more readable - I already tried it.